### PR TITLE
Vb 2717 implement non association check if prisoner and non association prison has a booking conflict

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

adds logic for handling non association events.
## What is the intent behind these changes?

Code changes for VB-2717 - handle non association events.

### CODE REVIEW

**Review by AledWEvans — COMMENTED** [review]
Some code refactoring needed but logic is good

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/EventType.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package uk.gov.justice.digital.hmpps.visitscheduler.service
+
+enum class EventType(val description: String) {
+  NON_ASSOCIATION_EVENT("Non-association"),
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/EventType.kt]: Updated to NotificationEventType as mentioned. I have renamed the description to reviewType. It is actually the reviewType value that is needed to be shown on app insights as described on VB-1939

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      val fromDate = getNowOrFutureDateTimeAtStartOfDay(nonAssociationChangedNotification.validFromDate)
+      val primaryPrisonerVisits = getBookedVisits(nonAssociationChangedNotification.prisonerNumber, prisonerDetails?.prisonId, fromDate, toDate)
+      val nonAssociationPrisonerVisits = getBookedVisits(nonAssociationChangedNotification.nonAssociationPrisonerNumber, prisonerDetails?.prisonId, fromDate, toDate)
+
+      if (primaryPrisonerVisits.isNotEmpty() && nonAssociationPrisonerVisits.isNotEmpty()) {
+        overlappingVisits = getOverLappingVisits(primaryPrisonerVisits, nonAssociationPrisonerVisits)
+      }
+    }
+
+    overlappingVisits.forEach {
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt]: Refactored. Did not include the 28 days limit as we should find visits only within 28 days limit

**Review by AledWEvans — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    return overlappingVisits.toList()
+  }
+
+  fun getBookedVisits(prisonerNumber: String, prisonCode: String?, startDateTime: LocalDateTime, endDateTime: LocalDateTime?): List<VisitDto> {
+    val visitFilter = VisitFilter(
+      prisonerId = prisonerNumber,
+      prisonCode = prisonCode,
+      startDateTime = startDateTime,
+      endDateTime = endDateTime,
+      visitStatusList = listOf(VisitStatus.BOOKED),
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt:105]: Just BOOKED as RESERVED / CHANGING will not be affected as the existing non association check should handle this scenario. There is a minor chance of an EVENT being raised after a visit has been reserved and (the visits having a new non association with another prisoner on same date) but there is an issue if the visit is not BOOKED after being RESERVED.

**Pull request merged**